packages/components/src/DataList/components/DataListLayoutInternal/DataListHeader.tsx

```diff
@@ -0,0 +1,55 @@
+import React from "react";
+import { DataListLayoutInternal } from "./DataListLayoutInternal";
+import { Breakpoints } from "../../DataList.const";
+import styles from "../../DataList.css";
+import {
+  DataListHeader as DataListHeaderType,
+  DataListItemTypeFromHeader,
+  DataListObject,
+  DataListProps,
+} from "../../DataList.types";
+import { sortSizeProp } from "../../DataList.utils";
+import { DataListLayoutProps } from "../DataListLayout/DataListLayout";
+
+interface DataListHeaderProps<T extends DataListObject> {
+  layouts: React.ReactElement<DataListLayoutProps<T>>[] | undefined;
+  mediaMatches?: Record<Breakpoints, boolean>;
+  headerData?: DataListItemTypeFromHeader<T, DataListHeaderType<T>>;
+  headerVisibility: NonNullable<DataListProps<T>["headerVisibility"]>;
+}
+
+export function DataListHeader<T extends DataListObject>({
+  layouts,
+  mediaMatches,
+  headerData,
+  headerVisibility,
+}: DataListHeaderProps<T>) {
+  const matchingMediaQueries = Object.keys(mediaMatches || {}).filter(
+    (key): key is Breakpoints => !!mediaMatches?.[key as Breakpoints],
+  );
+  const sortedVisibleBreakpoints = sortSizeProp(matchingMediaQueries);
+
+  // Determines if the header should be visible based on the headerVisibility
+  const showHeader = sortedVisibleBreakpoints.reduce(
+    (previousVisibility, breakpoint) => {
+      return headerVisibility[breakpoint] ?? previousVisibility;
+    },
+    true,
+  );
+
+  return (
+    <DataListLayoutInternal
+      layouts={layouts}
+      renderLayout={layout => {
+        return (
+          (showHeader && headerData && (
```

Comment:
This is why I like keeping the code organized, it makes it easy to spot something like this.

This check and how it returns a fragment if one of these disappears. should this be caught earlier? 

Don't render the `DataListLayoutInternal` if we don't want it to show (`showHeader`) and there's nothing to show (`headerData`) 🤔

Reply:
I added an early return for these conditionals.